File: code/015/python/main.py
"""From Python to Go: Python: 015 - Basic SSH."""

# Modules
import argparse
import datetime
import os
import sys
import time
import re
from dataclasses import dataclass
from typing import List
import yaml
import paramiko


# Classes
@dataclass
class Credentials:
    """Class to store credentials."""
    username: str
    password: str


@dataclass
class Result:
    """Class to store command execution data."""
    command: str
    output: str
    timestamp: datetime.datetime


class Device:
    """Class to interact with netowrk device."""
    def __init__(self, hostname: str, ip_address: str, credentials: Credentials):
        self.hostname = hostname
        self.ip_address = ip_address
        self.credentials = credentials

        self.results: List[Result] = []

    def execute_command(self, command: str) -> None:
        """Method to execute a command."""
        # Create a new SSH client
        client = paramiko.SSHClient()
        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

        # Connect to the device
        client.connect(
            self.ip_address,
            username=self.credentials.username,
            password=self.credentials.password,
            look_for_keys=False,
            allow_agent=False,
        )

        # Invoke the session
        session = client.invoke_shell()
        session.recv(65535)

        # Execute the command
        session.send(command + "\n")
        output = ""
        regex = re.compile(rf"{self.hostname}[#>]", flags=re.MULTILINE)
        print(f"{regex=}")
        while not regex.search(output):
            time.sleep(.1)
            output += session.recv(65535).decode("utf-8")
            print(output)

        # Store the result
        self.results.append(Result(command, output, datetime.datetime.now()))

        # Close the connection
        session.close()


# Functions
def read_args() -> argparse.Namespace:
    """Helper function to read CLI arguments."""
    parser = argparse.ArgumentParser(description="User input.")
    parser.add_argument("-i", "--inventory", type=str, help="Path to inventory file.")
    return parser.parse_args()


def load_inventory(filename: str, credentials: Credentials) -> List[Device]:
    """Function to load inventory data."""
    # Open file
    try:
        with open(filename, "r", encoding="utf-8") as f:
            data = yaml.safe_load(f)

    except FileNotFoundError as e:
        print(e)
        sys.exit(1)

    # Populate list of devices
    result = []
    for device in data:
        result.append(Device(credentials=credentials, **device))

    return result


def get_credentials() -> Credentials:
    """Function to get credentials."""
    username = os.getenv("AUTOMATION_USER")
    password = os.getenv("AUTOMATION_PASS")
    return Credentials(username, password)


# Main code
if __name__ == "__main__":
    # Read CLI arguments
    args = read_args()

    # Get credentials
    credentials = get_credentials()

    # Load inventory
    devices = load_inventory(args.inventory, credentials=credentials)

    # Execute command
    for device in devices:
        device.execute_command("show version")

    # Print results
    for device in devices:
        print(f"Device: {device.hostname}")
        for result in device.results:
            print(f"Command: {result.command}", f"Output: {result.output}", f"Timestamp: {result.timestamp}", sep="\n")

File: code/015/go/main.go
/* From Python to Go: Go: 015 - Basic SSH. */

package main

import (
	"bytes"
	"flag"
	"fmt"
	"log"
	"os"
	"time"

	"golang.org/x/crypto/ssh"
	"gopkg.in/yaml.v3"
)

// Imports

// Types and Receivers
type Arguments struct {
	/* Class to starte CLI arguments */
	Inventory string
}
type Crendetials struct {
	/* Struct to store credentials. */
	Username string
	Password string
}

type Result struct {
	/* Struct to store command execution result. */
	Command   string
	Output    string
	Timestamp time.Time
}

type Device struct {
	/* Struct to interact with netowrk device. */
	Hostname    string `yaml:"hostname"`
	IpAddress   string `yaml:"ip_address"`
	Crendetials Crendetials
	Result      []Result
}

func (d *Device) executeCommand(c string) {
	/* Method to execute command */
	interactiveAuth := ssh.KeyboardInteractive(
		func(user, instruction string, questions []string, echos []bool) ([]string, error) {
			answers := make([]string, len(questions))
			for i := range answers {
				answers[i] = (*d).Crendetials.Password
			}

			return answers, nil
		},
	)

	// Create a new SSH client
	sshClientConfig := &ssh.ClientConfig{
		User:            (*d).Crendetials.Username,
		Auth:            []ssh.AuthMethod{interactiveAuth},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
	}
	sshClient, err := ssh.Dial("tcp", fmt.Sprintf("%v:22", (*d).IpAddress), sshClientConfig)
	if err != nil {
		log.Fatalln("Failed to dial: ", err)
	}
	defer sshClient.Close()

	// Create session
	session, err := sshClient.NewSession()
	if err != nil {
		log.Fatalln("Failed to open the session: ", err)
	}
	defer session.Close()

	// Execute the command
	buffer := bytes.Buffer{}
	session.Stdout = &buffer
	if err := session.Run(c); err != nil {
		log.Fatalln("Failed to execute command: ", err)
	}

	// Update the result
	(*d).Result = append((*d).Result, Result{
		Command:   c,
		Output:    buffer.String(),
		Timestamp: time.Now(),
	})
}

// Functions
func readArgs() Arguments {
	/* Helper function to read CLI arguments */
	result := Arguments{}

	flag.StringVar(&result.Inventory, "i", "", "Path to the inventory file")

	flag.Parse()

	return result
}

func loadInventory(p string) *[]Device {
	/* Function to load inventory data. */

	// Open file
	bs, err := os.ReadFile(p)
	if err != nil {
		fmt.Println("Get error ", err)
		os.Exit(1)
	}

	// Load inventory
	result := &[]Device{}

	err = yaml.Unmarshal(bs, result)
	if err != nil {
		fmt.Println("Get error ", err)
		os.Exit(1)
	}

	// Return result
	return result
}

func getCredentials() Crendetials {
	/* Function to get credentials. */
	return Crendetials{
		Username: os.Getenv("AUTOMATION_USER"),
		Password: os.Getenv("AUTOMATION_PASS"),
	}
}

// Main
func main() {
	/* Core logic */
	// Read CLI arguments
	cliArgs := readArgs()

	// Get credentials
	sshCreds := getCredentials()

	// Load inventory
	inventory := loadInventory(cliArgs.Inventory)

	// Execute commands
	for i := 0; i < len(*inventory); i++ {
		(*inventory)[i].Crendetials = sshCreds
		(*inventory)[i].executeCommand("show version")
	}

	// Print results
	for i := 0; i < len(*inventory); i++ {
		for j := 0; j < len((*inventory)[i].Result); j++ {
			fmt.Printf(
				"Command: %v\nOutput: %v\nTimestamp: %v\n",
				(*inventory)[i].Result[j].Command,
				(*inventory)[i].Result[j].Output,
				(*inventory)[i].Result[j].Timestamp,
			)
		}
	}
}
